src/mesh.cpp

```diff
@@ -236,6 +236,63 @@ vector<Mesh::MaterialVolume> Mesh::material_volumes(
   return result;
 }
 
+void Mesh::sum_distance_xs(const Position& start, const Position& end) const
+{
+  Position direction = {end.x - start.x, end.y - start.y, end.z - start.z};
+  double magnitude =
+    std::sqrt(direction.x * direction.x + direction.y * direction.y +
+              direction.z * direction.z);
+
+  // Normalize the direction
+  direction.x /= magnitude;
+  direction.y /= magnitude;
+  direction.z /= magnitude;
+
+  Particle p;
+  SourceSite site;
+  site.E = 1.0;
+  site.particle = ParticleType::neutron;
+  site.r = start;
+  site.u = direction;
+  p.from_source(&site);
+
+  double total_distance_xs = 0.0;
+
+  while (true) {
+    Position r0 = p.r();
+
+    // Find distance to the next boundary and move
+    BoundaryInfo boundary = distance_to_boundary(p);
+    double distance = boundary.distance;
+    p.move_distance(distance);
+
+    double xs = 0.0;
+    if (p.material() != C_NONE) {
+      const Material& material = *model::materials[p.material()];
+      // Update internal state in Particle p
+      material.calculate_xs(p);
+      xs = p.macro_xs().total;
+    } else {
+      // Use default macro cross-section if no material
+      xs = p.macro_xs().total;
+    }
```

Comment:
I don't know how the material and cross-section interfaces work, but looking at your algorithm, it seems like you do the same last line regardless of the logic path, so you could do this:
```suggestion
    if (p.material() != C_NONE) {
      const Material& material = *model::materials[p.material()];
      // Update internal state in Particle p
      material.calculate_xs(p);
    } 
    double xs = p.macro_xs().total;
```